benchmarks/benchmark_serving.py

```diff
@@ -0,0 +1,256 @@
+            continue
+        filtered_dataset.append((prompt, prompt_len, output_len))
+
+    # Sample the requests.
+    sampled_requests = random.sample(filtered_dataset, num_requests)
+    return sampled_requests
+
+
+async def get_request(
+    input_requests: List[Tuple[str, int, int]],
+    request_rate: float,
+) -> AsyncGenerator[Tuple[str, int, int], None]:
+    for request in input_requests:
+        yield request
+
+        if request_rate == float("inf"):
+            # If the request rate is infinity, then we don't need to wait.
+            continue
+        # Sample the request interval from the exponential distribution.
+        interval = np.random.exponential(1.0 / request_rate)
+        # The next request will be sent after the interval.
+        await asyncio.sleep(interval)
+
+
+async def send_request(
+    api_url: str,
+    prompt: str,
+    prompt_len: int,
+    output_len: int,
+    config: dict,
+    progress_bar: tqdm = None,
+) -> None:
+    request_start_time = time.perf_counter()
+
+    headers = {"User-Agent": "Benchmark Client"}
+
+    # Use sample output_len if max_new_tokens not specified
+    if "max_new_tokens" in config:
+        output_len = config["max_new_tokens"]
+    else:
+        config["max_tokens"] = output_len
+
+    pload = {
+        "text": prompt,
+        "config": config,
+        "stream": False,
+    }
+
+    timeout = aiohttp.ClientTimeout(total=3 * 3600)
+    async with aiohttp.ClientSession(timeout=timeout) as session:
+        while True:
+            async with session.post(api_url, headers=headers, json=pload) as response:
+                chunks = []
+                async for chunk, _ in response.content.iter_chunks():
+                    chunks.append(chunk)
+            output = b"".join(chunks).decode("utf-8")
+            print(f"\n=== Request =====\n{pload}\n================\n")
+            print(f"=== Response ===\n{output}\n================\n")
+            if progress_bar:
+                progress_bar.update()
+            break
+
+    request_end_time = time.perf_counter()
+    request_latency = request_end_time - request_start_time
+    REQUEST_LATENCY.append((prompt_len, output_len, request_latency))
+
+
+async def benchmark(
+    api_url: str,
+    input_requests: List[Tuple[str, int, int]],
+    request_rate: float,
+    config: dict,
+    progress: bool,
+) -> None:
+    tasks: List[asyncio.Task] = []
+    progress_bar = tqdm(total=len(input_requests)) if progress else None
+    async for request in get_request(input_requests, request_rate):
+        prompt, prompt_len, output_len = request
+        task = asyncio.create_task(
+            send_request(api_url, prompt, prompt_len, output_len, config, progress_bar)
+        )
+        tasks.append(task)
+    await asyncio.gather(*tasks)
+
+
+def main(args: argparse.Namespace):
+    print(args)
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+
+    route_prefix = all_models[args.model_name].route_prefix
```

Comment:
If you start using command `python inference/serve.py --config_file <model_config_file.yaml> --simple --keep_serve_terminal` as described in the doc, the yaml file might not in the models directory. So we should use the `route_prefix` and `tokenizer_name_or_path` etc specified in the user conf file instead of `all_models[args.model_name]`, right?

Reply:
It currently only supports model from all_models. Yes, it should be added for more flexibility.